File: remove_housing.js
const fs = require('fs');
const { parser } = require('stream-json');
const { streamArray } = require('stream-json/streamers/StreamArray');
const { chain } = require('stream-chain');

const inputFile = 'data_20260603.json';
const tempFile = 'data_20260603_tmp.json';

console.log(`Aloitetaan suodatus RAM-ystävällisesti (Node.js): ${inputFile}`);

let processed = 0;
let removed = 0;
let first = true;

const terms = ["asunto oy", "asunto-osakeyhtiö", "kiinteistö oy", "kiinteistöosakeyhtiö", "as.oy", "as. oy", "asunto-oy"];

const pipeline = chain([
  fs.createReadStream(inputFile),
  parser(),
  streamArray(),
  data => {
    processed++;
    if (processed % 100000 === 0) {
      console.log(`Käsitelty ${processed} yritystä, poistettu asunto/kiinteistö-yhtiöitä tähän mennessä ${removed}...`);
    }

    const company = data.value;
    const names = company.names || [];
    const companyName = names.length > 0 ? (names[0].name || '') : '';
    const nameLower = companyName.toLowerCase();
    
    const mbl = company.mainBusinessLine;
    let industryCode = '';
    if (mbl && typeof mbl === 'object') {
        industryCode = mbl.type || '';
    } else if (typeof mbl === 'string') {
        industryCode = mbl;
    }
    
    let isHousing = false;
    for (const term of terms) {
      if (nameLower.includes(term)) {
        isHousing = true;
        break;
      }
    }
    
    if (!isHousing && industryCode.startsWith("682")) {
      isHousing = true;
    }
    
    if (isHousing) {
      removed++;
      return null;
    }
    return company;
  }
]);

const out = fs.createWriteStream(tempFile);
out.write('[\n');

pipeline.on('data', company => {
  if (!first) {
    out.write(',\n');
  }
  first = false;
  out.write(JSON.stringify(company));
});

pipeline.on('end', () => {
  out.write('\n]\n');
  out.end();
});

out.on('finish', () => {
  console.log(`\nSuodatus valmis!`);
  console.log(`Käsitelty yhteensä: ${processed} yritystä.`);
  console.log(`Poistettu lopullisesti: ${removed} asunto/kiinteistöyhtiötä.`);
  console.log(`Korvataan alkuperäinen tiedosto...`);
  
  fs.renameSync(tempFile, inputFile);
  console.log(`Tiedosto korvattu onnistuneesti! Voit poistaa remove_housing.js ja remove_housing.py tiedostot.`);
});

pipeline.on('error', (err) => {
  console.error('Virhe suodatuksessa:', err);
});

File: remove_housing.py
import json
import ijson
import os

def main():
    input_file = "data_20260603.json"
    temp_file = "data_20260603_tmp.json"
    
    print(f"Aloitetaan suodatus: {input_file} (tämä voi kestää hetken, tiedosto on iso)")
    
    processed_count = 0
    removed_count = 0
    
    try:
        with open(input_file, 'r', encoding='utf-8') as fin, open(temp_file, 'w', encoding='utf-8') as fout:
            fout.write("[\n")
            
            companies = ijson.items(fin, 'item')
            first = True
            
            for company in companies:
                processed_count += 1
                if processed_count % 100000 == 0:
                    print(f"Käsitelty {processed_count} yritystä, poistettu asunto/kiinteistö-yhtiöitä tähän mennessä {removed_count}...")
                
                names = company.get("names", [])
                company_name = names[0].get("name", "") if names else ""
                name_lower = company_name.lower()
                
                main_business_line = company.get("mainBusinessLine")
                industry_code = main_business_line.get("type", "") if isinstance(main_business_line, dict) else ""
                
                is_housing = False
                if any(term in name_lower for term in ["asunto oy", "asunto-osakeyhtiö", "kiinteistö oy", "kiinteistöosakeyhtiö", "as.oy", "as. oy", "asunto-oy"]):
                    is_housing = True
                elif industry_code and industry_code.startswith("682"):
                    is_housing = True
                    
                if is_housing:
                    removed_count += 1
                    continue
                    
                if not first:
                    fout.write(",\n")
                first = False
                
                json.dump(company, fout, ensure_ascii=False)
                
            fout.write("\n]\n")
            
        print(f"\nSuodatus valmis!")
        print(f"Käsitelty yhteensä: {processed_count} yritystä.")
        print(f"Poistettu lopullisesti: {removed_count} asunto/kiinteistöyhtiötä.")
        print("Korvataan alkuperäinen tiedosto uudella puhtaammalla versiolla...")
        
        os.replace(temp_file, input_file)
        print("Tiedosto korvattu onnistuneesti!")
        
    except Exception as e:
        print(f"Virhe: {e}")
        if os.path.exists(temp_file):
            os.remove(temp_file)

if __name__ == "__main__":
    main()
